### ISSUE TITLE
inverse_transform function for Imputer and HampelFilter

### ISSUE THREAD

**Issue by aiwalter** [alan-turing-institute/sktime#744]
**Is your feature request related to a problem? Please describe.**
I want to be able to use `Imputer` and `HampelFilter` inside `TransformedTargetForecaster` to be able to tune them together with a forecaster. However, this is currently not possible because there is not `inverse_transform()` function implemented. 

**Describe the solution you'd like**
I have checked the `sklearn.Imputer` and they also dont have an `inverse_tranform`. My idea however is to just implement a dummy `inverse_tranform` function that does nothing, so just returning input as output.

My thoughts on implementing this:
PRO
- enable tuning

CONTRA:
- inverting would not lead to the same original data (but we could store the original data inside the transformer to be able to do this at least for the insample transformations). Otherwise we could just raise a warning that inverting `Imputer` or `HampelFilter` is not unambiguously.

Would be interested in your opinion @mloning @fkiraly
___
Martin Walter <martin_friedrich.walter@daimler.com>, Mercedes-Benz AG on behalf of Daimler TSS GmbH.
https://github.com/Daimler/daimler-foss/blob/master/LEGAL_IMPRINT.md

**Comment by fkiraly** [comment]
Hm, yes - or we just want the effect of not having `inverse_transform` implemented to default to the identity transform?
We could do this inside `TransformedTargetForecaster`, or perhaps - better - in the abstract template class (whichever that is)?

**Comment by mloning** [comment]
Instead of doing the identity transform we could just skip them in the pipeline here:
https://github.com/alan-turing-institute/sktime/blob/bcce82cbabff00878d5347225d5679c8d30cf239/sktime/forecasting/compose/_pipeline.py#L130-L131
by adding something like this:
```python
 for _, _, transformer in self._iter_transformers(reverse=True): 
     if hasattr(transformer, "inverse_transform"):
         y_pred = transformer.inverse_transform(y_pred) 
```

Using the identity transform has the danger that for some transformers the inverse transform may simply not be implemented yet, even though it exists. 

I'm also not sure if we're missing something here. There seem to be some kinds of transforms where we do want to take the inverse, and others where we don't, e.g. for box-cox we do, for imputation we don't. Not sure if there's a better way of handling this.

**Comment by fkiraly** [comment]
I feel it's important here to try spell out a few example cases explicitly before we make a design decision - shall we do that? @aiwalter, would you mind explaining how you would like `Imputer` and `HampelFilter` to behave, specifically?

**Comment by aiwalter** [comment]
@fkiraly I have just seen that my previous statement partially incorrect:

> I have checked the` sklearn.Imputer` and they also dont have an `inverse_tranform`

This statement was correct, but `sklearn.Imputer` is depreciated and was replaced with `sklearn.SimpleImputer`, which now has an `inverse_transform` method. So my idea is to copy it ideally for `sktime.Imputer`. Also this might then work for `HampelFilter` as well.

> how you would like Imputer and HampelFilter to behave, specifically?

So the main purpose is to use the `fit_transform` method and to ideally tune the hyperparameters together with a forecaster and other transformers inside a `TransformedTargetForecaster`. Regarding `inverse_transform` I have only one requirement, which is to enhance tuning. The way we implement `inverse_transform` is for me thus not a particular requirement, but as stated I will try to copy the approach from `sklearn`, maybe I can even inherit sth.

___
Martin Walter <martin_friedrich.walter@daimler.com>, Mercedes-Benz AG on behalf of Daimler TSS GmbH.
https://github.com/Daimler/daimler-foss/blob/master/LEGAL_IMPRINT.md

**Comment by mloning** [comment]
The [`inverse_transform`](https://scikit-learn.org/stable/modules/generated/sklearn.impute.SimpleImputer.html#sklearn.impute.SimpleImputer.inverse_transform) of the `SimpleImputer` replaces the imputed values with np.nan again by remembering the indices of imputed values ("binary indicators"), those are basically what we would call annotations, i.e. label-timepoints pairs. 

This will work for out-of-sample predictions. In this case, the time points for imputed values in the training set will not be in the predicted horizon, so `inverse_transform` will be the identity transform. However, this will not work for in-sample predictions. Here, as some predicted values would be replaced with np.nan now - which is not want we want, right?

The key difference is that scikit-learn, as far as I'm aware, doesn't deal with missing values in the target variable. This simply raises an error:
```
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.impute import SimpleImputer
from sklearn.compose import TransformedTargetRegressor
tt = TransformedTargetRegressor(regressor=LinearRegression(), 
                                transformer=SimpleImputer())
X = np.arange(4).reshape(-1, 1)
y = np.exp(2 * X).ravel()
y[2] = np.nan
tt.fit(X, y)
tt.predict(X)
```

**Comment by fkiraly** [comment]
Hm, @mloning, @aiwalter, what would be an example where we would want `inverse_transform` of an imputer to *remove* imputed values again, as opposed to it being the identity transform, and why?
I can't construct an example where this would be beneficial or required by clean practice, but perhaps I just don't see it.

**Comment by aiwalter** [comment]
@fkiraly yes so basically why can we not just do the identity transform and perhaps raise a warning? Any ideas how `inverse_transform` of `HampelFilter` could look like?
So, what should we do now with this issue? 
___
Martin Walter <martin_friedrich.walter@daimler.com>, Mercedes-Benz AG on behalf of Daimler TSS GmbH.
https://github.com/Daimler/daimler-foss/blob/master/LEGAL_IMPRINT.md

**Comment by mloning** [comment]
Some of the options: 
1. create an estimator tag "skip-inverse-transform", check in the pipeline if tag is present and skip if not, e.g. `has_tag(transformer, "skip-inverse-transform")`,
1. check in the pipeline if inverse transform is available and skip if not, e.g. `hasattr(transformer, "inverse_transform")`,
1. implement the inverse transform as the identity transform,

Any other options? 

Out of the above options, I think I prefer 1 or 2. 1 is most explicit. 2 would require us to make sure that all transformers which in principle have an inverse transform but which hasn't been implemented yet to raise a `NotImplementedError` to avoid silent errors. 3 may not be intuitive, it's not what I would expect from calling `inverse_transform`.
